src/transaction/log_replication.cpp

```diff
@@ -1,173 +1,264 @@
-  {
-    std::unique_lock<std::mutex> ulock (m_redo_mutex);
-    m_redo_condvar.wait (ulock, [this]
-    {
-      return m_redo_lsa >= log_Gl.append.get_nxio_lsa ();
-    });
-  }
-} // namespace cublog
+/*
+ * Copyright 2008 Search Solution Corporation
+ * Copyright 2016 CUBRID Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+#include "log_replication.hpp"
+
+#include "log_impl.h"
+#include "log_recovery.h"
+#include "log_recovery_redo.hpp"
+#include "log_recovery_redo_parallel.hpp"
+#include "recovery.h"
+#include "thread_looper.hpp"
+#include "thread_manager.hpp"
+#include "transaction_global.hpp"
+
+#include <cassert>
+#include <chrono>
+#include <functional>
+
+namespace cublog
+{
+
+  replicator::replicator (const log_lsa &start_redo_lsa)
+    : m_redo_lsa { start_redo_lsa }
+  {
+    log_zip_realloc_if_needed (m_undo_unzip, LOGAREA_SIZE);
+    log_zip_realloc_if_needed (m_redo_unzip, LOGAREA_SIZE);
+
+    // depending on parameter, instantiate the mechanism to execute replication in parallel
+    // mandatory to initialize before daemon such that:
+    //  - race conditions, when daemon comes online, are avoided
+    //  - (even making abstraction of the race conditions) no log records are needlessly
+    //    processed synchronously
+    const int replication_parallel
+      = prm_get_integer_value (PRM_ID_PAGE_SERVER_REPLICATION_PARALLEL_COUNT);
+    assert (replication_parallel >= 0);
+    if (replication_parallel > 0)
+      {
+	m_parallel_replication_redo.reset (new cublog::redo_parallel (replication_parallel));
+      }
+
+    // Create the daemon
+    cubthread::looper loop (std::chrono::milliseconds (1));   // don't spin when there is no new log, wait a bit
+    auto func_exec = std::bind (&replicator::redo_upto_nxio_lsa, std::ref (*this), std::placeholders::_1);
+
+    if (replication_parallel > 0)
+      {
+        auto func_retire = std::bind (&replicator::wait_parallel_replication_idle, std::ref (*this));
+        m_daemon_task.reset (
+          new cubthread::entry_callable_task (std::move (func_exec), std::move (func_retire)));
+      }
+    else
+      {
+        // do not move task ownershit to itself - aka, do not let the task delete itself;
+        // it can be done, but prefer to do this locally for uniformity with the 'if' case
+        m_daemon_task.reset (
+          new cubthread::entry_callable_task (std::move (func_exec), false));
+      }

```

Comment:
`wait_parallel_replication_idle()` can be called during retire in synchronous case too, and for now it would do nothing. It would help to add an else clause to the `if (m_parallel_replication_redo != nullptr)` condition and a comment explaining that in sync case nothing is required. Of course, the name of wait_parallel_replication_idle() 
 would have to be changed too :).
```suggestion
    auto func_retire = std::bind (&replicator::wait_parallel_replication_idle, std::ref (*this));
    m_daemon_task.reset (new cubthread::entry_callable_task (std::move (func_exec), std::move (func_retire)));
```

Reply:
Initially it was like you suggest. Then I thought it is more explicit to only initialize what is needed when is needed.
Changed as suggested.